### PR TITLE
CoreFX FileSystemWatcher (Mac)

### PR DESCRIPTION
**Pull request by alexischr**
This is safe to merge, it only changes behavior for MacOS and there is a runtime fallback. There are 3 failing tests, all of which are new and will be disabled if not fixed.

* The bridge class is needed to preserve some useful Mono FSW backend implementations, like `KeventWatcher` which is the only option on FreeBSD, and  `DefaultWatcher` which is a working managed implementation. We can convert them from the custom interface to partial class implementations also in the future, and upstream to CoreFX.

* The Mono backend had some fixes pointed to by the CoreFX tests

#6966

### CODE REVIEW

**Comment by marek-safar** [comment]
@alexischr as discussed this is not exactly the approach we want to take, please wait for @luhenry to land the initial scaffolding to follow it up

**Comment by alexischr** [comment]
@marek-safar The work in this PR is not really System.Native related, I just imported two trivial functions from it so I can test

**Comment by alexischr** [comment]
build pkg

**Comment by alexischr** [comment]
build pkg

**Comment by alexischr** [comment]
@monojenkins build

**Comment by alexischr** [comment]
@monojenkins build

**Comment by alexischr** [comment]
build

File: data/config.in [hunk 1]
```diff
@@ -40,4 +40,5 @@
 	<dllmap dll="gdiplus.dll" target="@libgdiplus_install_loc@"  os="!windows"/>
 	<dllmap dll="gdi32" target="@libgdiplus_install_loc@" os="!windows"/>
 	<dllmap dll="gdi32.dll" target="@libgdiplus_install_loc@" os="!windows"/>
+	<dllmap dll="System.Native" target="__Internal"/>
```
Comment (marek-safar) [inline data/config.in]: @luhenry do we need it for the compiled in code?

File: mcs/class/System/net_4_x_System.dll.sources [hunk 2]
```diff
@@ -143,6 +143,24 @@ System.IO/InotifyWatcher.cs
 System.IO/KeventWatcher.cs
 System.IO/SearchPattern.cs
 System.IO/WindowsWatcher.cs
+System.IO/CoreFXFileSystemWatcherProxy.cs
+../../../external/corefx/src/System.IO.FileSystem.Watcher/src/System/IO/PatternMatcher.cs
+../../../external/corefx/src/Common/src/Interop/Windows/Interop.Errors.cs
```
Comment (marek-safar) [inline mcs/class/System/net_4_x_System.dll.sources]: This should go to windows .sources only and similarly for other Unix/OSX/Windows .cs files

File: mcs/class/System/net_4_x_System.dll.sources [hunk 3]
```diff
@@ -143,6 +143,24 @@ System.IO/InotifyWatcher.cs
...
+../../../external/corefx/src/Common/src/Interop/Unix/Interop.Errors.cs
+../../../external/corefx/src/Common/src/Interop/OSX/Interop.EventStream.cs
+../../../external/corefx/src/Common/src/Interop/Unix/Interop.IOErrors.cs
+../../../external/corefx/src/Common/src/Interop/OSX/Interop.RunLoop.cs
+../../../external/corefx/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs
+../../../external/corefx/src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs
```
Comment (marek-safar) [inline mcs/class/System/net_4_x_System.dll.sources]: This should go to MacOS .sources only

**Comment by marek-safar** [comment]
build

File: .gitmodules [hunk 4]
```diff
@@ -36,7 +36,7 @@
 	branch = mono
 [submodule "external/corefx"]
 	path = external/corefx
-	url = git://github.com/mono/corefx.git
+	url = git@github.com:mono/corefx.git
```
Comment (alexischr) [inline .gitmodules]: I'll fix this before merging

**Comment by marek-safar** [comment]
xammac_net_4_5 profile fails to build

**Comment by marek-safar** [comment]
@alexischr I had to revert it as it breaks when building with master corefx submodule. Could you please rebase it on top of master

**Comment by akoeplinger** [comment]
I fixed the issue and pushed to master, should be good now :)

**Pull request merged**